HW_1.py

```diff
@@ -0,0 +1,440 @@
+# + Данные классы не должны иметь <ins>публичных методов</ins> `complement` и метода для проверки алфавита, так как они уже должны быть реализованы в `NucleicAcidSequence`.
+# 
+# Напишите класс `AminoAcidSequence`:
+# + Данный класс реализует интерфейс `BiologicalSequence`
+# + Добавьте этому классу один любой метод, подходящий по смыслу к аминокислотной последовательности. Например, метод для нахождения изоэлектрической точки, молекулярного веса и т.д.
+# 
+# Комментарий по поводу метода `NucleicAcidSequence.complement`, так как я хочу, чтобы вы сделали его опредедённым образом:
+# 
+# При вызове `dna.complement()` или условного `dna.check_alphabet()` должны будут вызываться соответствующие методы из `NucleicAcidSequence`. При этом, данный метод должен обладать свойством полиморфизма, иначе говоря, внутри `complement` не надо делать условия а-ля `if seuqence_type == "DNA": return self.complement_dna()`, это крайне не гибко. Данный метод должен опираться на какой-то общий интерфейс между ДНК и РНК. Создание экземпляров `NucleicAcidSequence` не подразумевается, поэтому код `NucleicAcidSequence("ATGC").complement()` не обязан работать, а в идеале должен кидать исключение `NotImplementedError` при вызове от экземпляра `NucleicAcidSequence`
+# 
+# Вся сложность задания в том, чтобы правильно организовать код. Если у вас есть повторяющийся код в сестринских классах или родительском и дочернем, значит вы что-то делаете не так.
+# 
+# 
+# Маленькое замечание: По-хорошему, между классом `BiologicalSequence` и классами `NucleicAcidSequence` и `AminoAcidSequence`, ещё должен быть класс-прослойка, частично реализующий интерфейс `BiologicalSequence`, но его писать не обязательно, так как задание и так довольно большое (правда из-за этого у вас неминуемо возникнет повторяющийся код в классах `NucleicAcidSequence` и `AminoAcidSequence`)
+
+# In[73]:
+
+
+from abc import ABC, abstractmethod
+class BiologicalSequence(ABC):
+    def __init__(self, bioseq):
+        self.bioseq = bioseq
+        
+    def __len__(self):
+        return len(bioseq)
+
+    def __getitem__(self, idxs):
+        return self.bioseq[idxs]
+
+    def check_alphabet(self):
+        nuclds = ['A', 'T', 'G', 'C', 'U']
+        for letter in self.bioseq:
+            if letter not in nuclds:
+                return "Oops, that's not a biological sequence! Try again, please (use only ATGCU)"
+            
+class NucleicAcidSequence(BiologicalSequence):
+    def __init__(self, bioseq):
+        super().__init__(bioseq)
+        
+    def complement(self):
+        if type(self) == NucleicAcidSequence:
+            raise NotImplementedError('A wrong class or a wrong method?')
+        compl_pairs = {'A': 'T', 
+                       'G': 'C', 
+                       'T': 'A', 
+                       'C': 'G', 
+                       'U': 'A'}
+        compl_res = ''
+        for nucl in self.bioseq:
+            compl_res += compl_pairs[nucl]
+        return compl_res
```

Comment:
1. Представь ситуацию, что у нас есть 15 различных нуклеиновых кислот помимо ДНК и РНК и у них есть свои классы и, более того, у каждой из них свои отдельные таблицы для комплементирования, как тогда удобнее организовать код? Можно задать для каждого класса свои таблицы и хранить их в каком-то атрибуте (допустим, `complement_table`). В таком случае здесь нам не придётся писать отдельные условия под каждую НК, так как будет использован общий для всех типов интерфейс
2. Когда мы делаем комплементарную ДНК, мы ожидаем получить на выходе ДНК, с РНК аналогично, а у тебя получаются строки)

```suggestion
    def complement(self):
        if type(self) == NucleicAcidSequence:
            raise NotImplementedError('A wrong class or a wrong method?') 
        compl_res = ''
        for nucl in self.bioseq:
            compl_res += self.complement_table[nucl]
        return type(self)(compl_res)
```